pkg/controller/endpointslice/reconciler.go

```diff
@@ -0,0 +1,288 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package endpointslice
+
+import (
+	"fmt"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	discovery "k8s.io/api/discovery/v1alpha1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/sets"
+	clientset "k8s.io/client-go/kubernetes"
+	endpointutil "k8s.io/kubernetes/pkg/controller/util/endpoint"
+)
+
+// Reconciler is responsible for transforming current EndpointSlice state into desired state
+type Reconciler struct {
+	client               clientset.Interface
+	service              corev1.Service
+	maxEndpointsPerSlice int32
+}
+
+// EndpointMeta includes the attributes we group slices on, this type helps with that logic in reconciler
+type EndpointMeta struct {
+	Ports       []discovery.EndpointPort `json:"ports" protobuf:"bytes,2,rep,name=ports"`
+	AddressType *discovery.AddressType   `json:"addressType" protobuf:"bytes,3,rep,name=addressType"`
+}
+
+// reconcile takes a set of pods currently matching a service selector and compares them with the
+// endpoints already present in any existing endpoint slices for the given service. It creates, updates,
+// or deletes endpoint slices to ensure the desired set of pods are represented by endpoint slices.
+func (r *Reconciler) reconcile(pods []*corev1.Pod, nodes []*corev1.Node, existingSlices []*discovery.EndpointSlice, triggerTime time.Time) error {
+	// Build data structures for existing state
+	existingSlicesByPortHash := map[string][]discovery.EndpointSlice{}
+	existingPortHashes := sets.String{}
+	for _, existingSlice := range existingSlices {
+		epHash := endpointPortsToHash(existingSlice.Ports)
+		existingSlicesByPortHash[epHash] = append(existingSlicesByPortHash[epHash], *existingSlice)
+		existingPortHashes.Insert(epHash)
+	}
+	nodesByName := map[string]*corev1.Node{}
```

Comment:
pass the NodeLister into the Reconciler?  No need to reconstruct the cache here.

Reply:
or have a interface GetNodeByName()

have an adapter for the real NodeLister
have another fake implementation of GetNodeByName() 


or have a initializer for the fake NodeLister in unit tests.

Reply:
nodeLister ended up being straightforward enough to use here, good call